[1.x] feat: add invalid-indexes error checking to the advance-runner

renan061 opened the pull request.

renan061 committed:
feat(advance-runner): detect invalid claim indexes

marcelstanley commented on offchain/advance-runner/src/broker.rs:
```diff
@@ -98,47 +105,63 @@ impl BrokerFacade {
+        let last_claim_event = self
             .client
-            .peek_latest(&self.claims_stream)
+            .read_latest(&self.claims_stream)
```
I think we should keep the old name here:
```suggestion
            .peek_latest(&self.claims_stream)
```

marcelstanley commented on offchain/rollups-events/src/broker/mod.rs:
```diff
@@ -155,15 +151,15 @@ impl Broker {
     /// This function doesn't block; if there is no event in the stream it returns None.
     #[tracing::instrument(level = "trace", skip_all)]
-    pub async fn peek_latest<S: BrokerStream>(
+    pub async fn read_latest<S: BrokerStream>(
```
Since we're just checking what is the latest event, I'd rather keep the original name:
```suggestion
    pub async fn peek_latest<S: BrokerStream>(
```

marcelstanley commented on offchain/rollups-events/tests/integration.rs:
```diff
@@ -111,22 +111,22 @@ async fn test_it_produces_events() {
 
 #[test_log::test(tokio::test)]
-async fn test_it_peeks_in_stream_with_no_events() {
+async fn test_it_reads_latest_in_stream_with_no_events() {
```
I'd revert to using `peek` wherever it was changed.

marcelstanley commented on offchain/indexer/tests/integration.rs:
```diff
@@ -38,7 +38,7 @@ async fn indexer_inserts_inputs() {
     let mut inputs = vec![];
     for i in 0..N {
-        let input = state.produce_input_in_broker(i).await;
+        let input = state.add_input_in_broker(i).await;
```
Perhaps replacing `in_broker` with `to_stream` in all similar named functions would make more sense.
```suggestion
        let input = state.add_input_to_stream(i).await;
```

marcelstanley reviewed (commented):
I believe this PR should come after #551, so some rebasing will be needed.

Also, I think we should improve the commit message for 2731d76c0e0b1e75b93e1869f46ead5d6699c833, explaining in more details what was done.

GMKrieger committed:
chore: improve logging

marcelstanley reviewed (commented):
Can you please change ab96dd6c22f421e28d938c7aa0271fa3bb3484d4 commit message to `feat(advance-runner): detect invalid claim input indexes`?

vfusco commented on offchain/dispatcher/src/drivers/context.rs:
```diff
@@ -65,6 +65,13 @@ impl Context {
         broker: &impl BrokerSend,
     ) -> Result<(), BrokerFacadeError> {
         let input_block_number = input.block_added.number.as_u64();
+        let input_epoch = self.calculate_epoch(input_block_number);
+        self.last_finished_epoch.map(|last_finished_epoch| {
+            // Asserting that the calculated epoch comes after the last finished epoch.
+            // (If last_finished_epoch == None then we don't need the assertion.)
+            assert!(input_epoch > last_finished_epoch)
+        });
+
```
Agree, this is a fix to assert the input epoch correctness before `finish_epoch_if_needed` and `enqueue_input`.

vfusco commented on offchain/advance-runner/src/broker.rs:
```diff
@@ -101,24 +108,40 @@ impl BrokerFacade {
-                tracing::trace!(?event, "got last claim produced");
-                rollups_claim.epoch_index <= event.payload.epoch_index
+                let last_claim = event.payload;
+                tracing::trace!(?last_claim, "got last claim from Redis");
```
from broker stream

vfusco reviewed (changes requested):
Some small changes only

GMKrieger committed:
fix(dispatcher): check epoch before finishing

This asserts the input epoch correctness before finish_epoch_if_needed and enqueue_input.

marcelstanley reviewed (dismissed):
Approved, but I'd rather amending commit `fix(dispatcher): check epoch before finishing` to
```
fix(dispatcher): check epoch before finishing

This asserts the input epoch correctness before finish_epoch_if_needed and enqueue_input.
```
